File: tests/RenderScriptTests/SceneGraph/src/com/android/scenegraph/Transform.java
/*
 * Copyright (C) 2011 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.scenegraph;

import java.lang.Math;
import java.util.ArrayList;

import android.renderscript.*;
import android.renderscript.Matrix4f;
import android.util.Log;

/**
 * @hide
 */
public abstract class Transform extends SceneGraphBase {
    Transform mParent;
    ArrayList<Transform> mChildren;

    ScriptField_SgTransform mField;
    ScriptField_SgTransform.Item mTransformData;

    public Transform() {
        mChildren = new ArrayList<Transform>();
        mParent = null;
    }

    public void appendChild(Transform t) {
        mChildren.add(t);
        t.mParent = this;
        updateRSChildData(true);
    }

    abstract void initLocalData();

    void updateRSChildData(boolean copyData) {
        if (mField == null) {
            return;
        }
        RenderScriptGL rs = SceneManager.getRS();
        if (mChildren.size() != 0) {
            Allocation childRSData = Allocation.createSized(rs, Element.ALLOCATION(rs),
                                                            mChildren.size());
            mTransformData.children = childRSData;

            Allocation[] childrenAllocs = new Allocation[mChildren.size()];
            for (int i = 0; i < mChildren.size(); i ++) {
                Transform child = mChildren.get(i);
                childrenAllocs[i] = child.getRSData().getAllocation();
            }
            childRSData.copyFrom(childrenAllocs);
        }
        if (copyData) {
            mField.set(mTransformData, 0, true);
        }
    }

    ScriptField_SgTransform getRSData() {
        if (mField != null) {
            return mField;
        }

        RenderScriptGL rs = SceneManager.getRS();
        if (rs == null) {
            return null;
        }
        mField = new ScriptField_SgTransform(rs, 1);

        mTransformData = new ScriptField_SgTransform.Item();
        mTransformData.name = getNameAlloc(rs);
        mTransformData.isDirty = 1;
        mTransformData.timestamp = 1;

        initLocalData();
        updateRSChildData(false);

        mField.set(mTransformData, 0, true);
        return mField;
    }
}

File: tests/RenderScriptTests/SceneGraph/src/com/android/scenegraph/transform.rs
// Copyright (C) 2011 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#pragma version(1)

#pragma rs java_package_name(com.android.modelviewer)

#include "scenegraph_objects.rsh"

rs_script gTransformScript;

typedef struct {
    int changed;
    rs_matrix4x4 *mat;
} ParentData;

//#define DEBUG_TRANSFORMS
static void debugTransform(SgTransform *data, const ParentData *parent) {
    rsDebug("****** <Transform> ******", (int)data);
    printName(data->name);
    rsDebug("isDirty", data->isDirty);
    rsDebug("parent", (int)parent);
    rsDebug("child ", rsIsObject(data->children));

    // Refresh matrices if dirty
    if (data->isDirty && rsIsObject(data->components)) {
        uint32_t numComponenets = rsAllocationGetDimX(data->components);
        for (int i = 0; i < numComponenets; i ++) {
            const SgTransformComponent *comp = NULL;
            comp = (const SgTransformComponent *)rsGetElementAt(data->components, i);

            if (rsIsObject(comp->name)) {
                rsDebug((const char*)rsGetElementAt(comp->name, 0), comp->value);
                rsDebug("Type", comp->type);
            } else {
                rsDebug("no name", comp->value);
                rsDebug("Type", comp->type);
            }
        }
    }

    rsDebug("timestamp", data->timestamp);
    rsDebug("****** </Transform> ******", (int)data);
}

static void appendTransformation(int type, float4 data, rs_matrix4x4 *mat) {
    rs_matrix4x4 temp;

    switch (type) {
    case TRANSFORM_TRANSLATE:
        rsMatrixLoadTranslate(&temp, data.x, data.y, data.z);
        break;
    case TRANSFORM_ROTATE:
        rsMatrixLoadRotate(&temp, data.w, data.x, data.y, data.z);
        break;
    case TRANSFORM_SCALE:
        rsMatrixLoadScale(&temp, data.x, data.y, data.z);
        break;
    }
    rsMatrixMultiply(mat, &temp);
}

void root(const rs_allocation *v_in, rs_allocation *v_out, const void *usrData) {

    SgTransform *data = (SgTransform *)rsGetElementAt(*v_in, 0);
    const ParentData *parent = (const ParentData *)usrData;

#ifdef DEBUG_TRANSFORMS
    debugTransform(data, parent);
#endif //DEBUG_TRANSFORMS

    rs_matrix4x4 *localMat = &data->localMat;
    rs_matrix4x4 *globalMat = &data->globalMat;

    // Refresh matrices if dirty
    if (data->isDirty && rsIsObject(data->components)) {
        bool resetLocal = false;
        uint32_t numComponenets = rsAllocationGetDimX(data->components);
        for (int i = 0; i < numComponenets; i ++) {
            if (!resetLocal) {
                // Reset our local matrix only for component transforms
                rsMatrixLoadIdentity(localMat);
                resetLocal = true;
            }
            const SgTransformComponent *comp = NULL;
            comp = (const SgTransformComponent *)rsGetElementAt(data->components, i);
            appendTransformation(comp->type, comp->value, localMat);
        }
    }

    if (parent) {
        data->isDirty = (parent->changed || data->isDirty) ? 1 : 0;
        if (data->isDirty) {
            rsMatrixLoad(globalMat, parent->mat);
            rsMatrixMultiply(globalMat, localMat);
        }
    } else if (data->isDirty) {
        rsMatrixLoad(globalMat, localMat);
    }

    ParentData toChild;
    toChild.changed = 0;
    toChild.mat = globalMat;

    if (data->isDirty) {
        toChild.changed = 1;
        data->timestamp ++;
    }

    if (rsIsObject(data->children)) {
        rs_allocation nullAlloc;
        rsForEach(gTransformScript, data->children, nullAlloc, &toChild, sizeof(toChild));
    }

    data->isDirty = 0;
}
